File: ruby/abc026_c.rb
n = gets.to_i
$p = Array.new(n+1){[]}
(n-1).times{|i|
	b = gets.to_i
	$p[b] << i+2
}

def dfs(i)
	salaries = []
	return 1 if $p[i].length==0
	$p[i].each do |j|
		salaries << dfs(j)
	end
	min_salary = salaries.min
	max_salary = salaries.max
	max_salary + min_salary + 1
end

p dfs(1)

File: python/abc-c/abc026_c.py
n = int(input())
p = [[] for _ in range(n + 1)]
for i in range(n - 1):
    b = int(input())
    p[b].append(i + 2)


def dfs(i):
    if not p[i]:
        return 1
    salaries = []
    for j in p[i]:
        salaries.append(dfs(j))
    return max(salaries) + min(salaries) + 1


print(dfs(1))
